src/strands/tools/registry.py

```diff
@@ -24,244 +24,318 @@
+            FileNotFoundError: If specified tool file doesn't exist.
         """
-        tool_config = {}
-        logger.debug("getting tool configurations")
+        if tool is None:
+            raise ValueError("Tool cannot be None")
 
-        # Add all registered tools
-        for tool_name, tool in self.registry.items():
-            # Make a deep copy to avoid modifying the original
-            spec = tool.tool_spec.copy()
-            try:
-                # Normalize the schema before validation
-                spec = normalize_tool_spec(spec)
-                self.validate_tool_spec(spec)
-                tool_config[tool_name] = spec
-                logger.debug("tool_name=<%s> | loaded tool config", tool_name)
-            except ValueError as e:
-                logger.warning("tool_name=<%s> | spec validation failed | %s", tool_name, e)
+        # Extract the tool name to check if it exists
+        tool_name = self._extract_tool_name(tool)
+        if not tool_name:
+            raise ValueError("Could not extract tool name from provided tool")
 
-        # Add any dynamic tools
-        for tool_name, tool in self.dynamic_tools.items():
-            if tool_name not in tool_config:
-                # Make a deep copy to avoid modifying the original
-                spec = tool.tool_spec.copy()
-                try:
-                    # Normalize the schema before validation
-                    spec = normalize_tool_spec(spec)
-                    self.validate_tool_spec(spec)
-                    tool_config[tool_name] = spec
-                    logger.debug("tool_name=<%s> | loaded dynamic tool config", tool_name)
-                except ValueError as e:
-                    logger.warning("tool_name=<%s> | dynamic tool spec validation failed | %s", tool_name, e)
+        # Check if the tool exists before making any changes
+        if tool_name not in self.agent_tools:
+            raise ValueError(f"Tool '{tool_name}' not found in registry")
 
-        logger.debug("tool_count=<%s> | tools configured", len(tool_config))
-        return tool_config
+        # Make a backup of the current tool in case the update fails
+        original_tool = self.agent_tools.get(tool_name)
 
-    # mypy has problems converting between DecoratedFunctionTool <-> AgentTool
-    def register_tool(self, tool: AgentTool) -> None:
-        """Register a tool function with the given name.
+        try:
+            # Update the tool
+            self._update_single_tool(tool, validate_unique=False)
+            return tool_name
+        except Exception as e:
+            # Restore the original tool if the update fails
+            if original_tool:
+                self.agent_tools[tool_name] = original_tool
+            logger.warning("tool_name=<%s> | failed to update tool | %s", tool_name, e)
+            raise ValueError(f"Failed to update tool '{tool_name}': {str(e)}") from e
+
+    def delete_tool(self, tool_name: str) -> str:
```

Comment:
Given that this class is named `ToolRegistry` perhaps we can name this as just `delete`?  Same with the other `{verb}_{noun}` names?